src/nfa/nfa.cc

```diff
@@ -193,21 +193,155 @@ void Nfa::add_trans(State stateFrom, Symbol symbolOnTransition, State stateTo) {
     auto transitionFromStateIter = transitionrelation[stateFrom].begin();
     for (; transitionFromStateIter != transitionrelation[stateFrom].end(); ++transitionFromStateIter) {
         if (transitionFromStateIter->symbol == symbolOnTransition) {
+            // Add transition with symbolOnTransition already used on transitions from stateFrom.
             transitionFromStateIter->states_to.insert(stateTo);
             return;
         } else if (transitionFromStateIter->symbol > symbolOnTransition) {
             break;
         }
     }
 
+    // Add transition to a new TransSymbolStates struct with symbolOnTransition yet unused on transitions from stateFrom.
     transitionrelation[stateFrom].insert(transitionFromStateIter, TransSymbolStates(symbolOnTransition, stateTo));
 }
 
+void Nfa::remove_trans(State src, Symbol symb, State tgt)
+{
+    if (!has_trans(src, symb, tgt))
+    {
+        throw std::invalid_argument(
+                "Transition [" + std::to_string(src) + ", " + std::to_string(symb) + ", " +
+                std::to_string(tgt) + "] does not exist.");
+    }
+
+    auto transitionFromStateIter{ transitionrelation[src].begin() };
+    while (transitionFromStateIter != transitionrelation[src].end())
+    {
+        if (transitionFromStateIter->symbol == symb)
+        {
+            transitionFromStateIter->states_to.remove(tgt);
+            if (transitionFromStateIter->states_to.empty())
+            {
+                transitionrelation[src].remove(*transitionFromStateIter);
+            }
+
+            return;
+        }
+
+        ++transitionFromStateIter;
+    }
+}
+
 State Nfa::add_new_state() {
     transitionrelation.emplace_back();
     return transitionrelation.size() - 1;
 }
 
+void Nfa::remove_epsilon(const Symbol epsilon)
+{
+    *this = Mata::Nfa::remove_epsilon(*this, epsilon);
```

Comment:
Is this safe? It feels like something, that could blow someones balls out. I mean, you are practically rewriting object from its own method.

Reply:
I will tag @ondrik here, since, he's proficient in C++.

Reply:
As I said below, I have seen this used before. However, I will take a closer look whether it can be a problem or not.

Reply:
I believe it should be the same as

```cpp
void remove_epsilon(Nfa* this, const Symbol epsilon)
{
    *this = Mata::Nfa::remove_epsilon(*this, epsilon);
```
With the assumption that `this != nullptr`. Do you find this piece of code strange?

Reply:
No I don't. If this is the case, that member functions are translated into such snippet, and they are compiled the same, then it is OK. However, if there is some additional code for member functions, then there could be problem. Since, after `*this = ...` where exactly are you? In new object? Old object? 

However, if Ondra is not worried, then I will trust his judgement.

Reply:
After some digging, I come up with the following:
- The code should be generally correct.
- There could be UB when some member variable of the newly created NFA is not initialized (not this case).
- Assigning like `this = &Nfa()` or `this = new Nfa()` would be dangerous, assigning like `*this = ` should be not.

Some references:
- https://stackoverflow.com/questions/3887849/using-this-in-c-class-method-to-fully-overwrite-self-instantiation
- https://stackoverflow.com/questions/13476879/how-can-you-assign-a-value-to-the-pointer-this-in-c

The method signature is basically as shown above by @ondrik.

Therefore, I do not see any dangers in using `*this =` in this context. However, if you find it mysterious in any ways, I can rewrite it as suggested earlier: assign to member variables from the newly created NFA instead.

Reply:
I think, it is OK to keep it like this. Well done on the research.

In case there is any UB, make a note that Ondrik will fix it.